### ISSUE TITLE
Append a random string to the filename to avoid attacks by guessing.

### ISSUE THREAD

**Issue by Gugli** [thunderbird/webdav-filelink#2]
I'm using this Webdav provider for Filelink, and it works great. Thanks for the good work.

For security reasons, I'm not comfortable having public files with easily guessable names (public_url/Documents.zip and such likes). 
I've taken on the habit to manually add a random string to my filename (e.g. "MjNm9uB55ln2ShABO53H_Documents.zip" ). But it is a bit of work every time I send something.

I'd really like the option to be included in this Thunderbird module.

Do you think it fits the intended purpose of your module ? If I were to make a pull request for that, do you think it will get merged ?

Thanks

**Comment by sprintman** [comment]
Sounds like a good idea. It would also prevent conflicts if the same filename was used twice (e.g. sending a newer version).
Instead of changing the filename, I would create random subfolder so that the filename doesn't get mangled.

**Comment by krumboeck** [comment]
I would also prefer random subfolders. Why this issue is still not fixed? What would be the price to get this issue fixed (no joke)?

**Comment by krumboeck** [comment]
Ping!

**Comment by FranseFrikandel** [comment]
Since this issue first got raised, it appears the repository has changed hands and is now maintained by the thunderbird team instead. Any chances this will still be added? I've been happily using my own modified version from the open PR. I really think this brings many advantages, as definetly not uncommon for me to have sent attachments with the same name which were different files, so them not being overridden is very useful to me.

**Comment by xpseudonym** [comment]
May be I'm missing something - but, from my experience, all that is necessary to present a webpage generated by webdav with a list of links to the files on the server is to simply enter only the domain part of a webdav file link.  So, however much obfuscation is done to the file name or directory structure, it can be all stripped away very easily.  So, for example, a link to `http://my.domain/mycomplicatedfilename` is revealed by going to `http://my.domain`
?

**Comment by FranseFrikandel** [comment]
My personal reason for this isn't really for security. Right now, if you happen to send a file with the same name of some file you sent earlier, it'll overwrite the initial file.

Regardless, because the extension allows you to seperate the private and public URL, you can disable webDAV completely on the public URL, and require authentication on the private URL.

**Comment by Gugli** [comment]
> all that is necessary to present a webpage generated by webdav with a list of links to the files on the server is to simply enter only the domain part of a webdav file link. 

This file-listing feature (see "Options +Indexes" in Apache) can be deactivated. Many Internet facing servers deactivate it, as it basically render all hosted files open to the public. To me, any server (or more precisely, VHosts) used for webdav-filelink should disable Indexes, or use authentication.
When Indexes are disabled, adding a random folder or prefix to the file acts somewhat like as a password, but does not require the recipient to enter any credentials, making it secure without being less convenient.

**Comment by xpseudonym** [comment]
> This file-listing feature (see "Options +Indexes" in Apache) can be deactivated. Many Internet facing servers deactivate it, as it...

Thanks @Gugli  - I'm using Hetzner's Storagebox service, I'll drop them a line and see if they can disable the Indexes.

**Comment by jobisoft** [comment]
I will add support for this. A couple of suggestions:/questions:

* we should make it optional in the management page for each configured provider, or not?
* the random folder should be created only once and then be re-used, or do you want to create a random folder for each upload?

**Comment by krumboeck** [comment]
The random folder should be created every upload. This prevents overriding a file with the same name.
I don't know if it makes sense with other providers or not. I can imagine that it doesn't make sense with every provider.

**Comment by FranseFrikandel** [comment]
Indeed, a new random folder is generated for each uploaded file. This has 2 main reasons:
 - Sending a file with the same filename again will not cause the original to be overwritten. Currently this is the case if you don't use the modified version from my PR.
 - Users cannot easily find files you sent to other people by guessing filenames, since every file is in another random folder that is hard to guess

The reason random folders are used instead of random filenames is so that the actual filename is preserved when sent.

**Comment by Gugli** [comment]
2 additional comments : 

- I'm not sure every webdav server allows for folder creation (or "collections" in HTTP parlance IIRC). For such cases, using a prefix in the filename could be and acceptable workaround. I don't think it is worth implementing that before the need arises, but if it does, an additional setting to toggle between the 2 methods should do the trick.

- For this to work, the prefix/folder name should use a robust random-genaration scheme. It probably doesn't need top-notch randomness, but it's a point that IMO deserves a double check. 

@FranseFrikandel I looked at the PR, and 
I'm not super convinced about `Math.random().toString(36).substr(2, 8);`
Don't you think it would be best to use : https://developer.mozilla.org/en-US/docs/Web/API/Crypto/getRandomValues ?
(eg. generating 20 bytes, and converting them to hexadecimal)

**Comment by jobisoft** [comment]
I recently added the connection check in the management page, which uses an OPTIONS request. If that response gives me information about the ability to create subfolders, then I could deactivate that feature if unsupported.

**Comment by FranseFrikandel** [comment]
@Gugli honestly, I just kinda hacked it in way back then and I didn't really know much JS at all (still not a lot tbf, more of a Python guy), and I also personally mostly found it very annoying files would be overridden and wasen't so much focused on the security aspect.

I do agree there isn't a single reason not to use the cryptographic random number generator, not entirely sure if you'd necessarily just convert it to hexadecimal, or perhaps just put the cryptographic numbers together into a bigint and base64 encode that again or something?

**Comment by celmi** [comment]
> I will add support for this. A couple of suggestions:/questions:
> 
>     * we should make it optional in the management page for each configured provider, or not?

IMHO it should be a default and it shouldn't even be possible to change the behaviour. I think the typical usecase is a thunderbird user sending various e-mails to various people who don't know each other and should be prevented from accessing attachments of other people. I cannot think of another usecase where the other option would be wanted (?)

> 
>     * the random folder should be created only once and then be re-used, or do you want to create a random folder for each upload?

it should always be a new random folder for each e-mail. So, let's say, one e-mail has three attachments, you can let them all upload to the same random folder.  But the next e-mail I create should upload large attachments to a new random folder.

**Comment by 3rd3** [comment]
I've submitted PR #18 based on this thread. It includes the discussed changes and is ready for review. Please let me know if any further improvements are needed.
